sprocs/variants_select_submission_for_manual_grading.sql

```diff
@@ -0,0 +1,34 @@
+DROP FUNCTION IF EXISTS variants_select_submission_for_manual_grading(bigint,boolean);
+DROP FUNCTION IF EXISTS variants_select_submission_for_manual_grading(bigint,bigint);
+
+CREATE OR REPLACE FUNCTION
+    variants_select_submission_for_manual_grading (
+        IN variant_id bigint,
+        IN check_submission_id bigint DEFAULT NULL
+    ) RETURNS TABLE (submission submissions)
+AS $$
+BEGIN
+    PERFORM variants_lock(variant_id);
+
+    -- start with the most recent submission
+    SELECT s.*
+    INTO submission
+    FROM submissions AS s
+    WHERE s.variant_id = variants_select_submission_for_manual_grading.variant_id
+    ORDER BY s.date DESC, s.id DESC
+    LIMIT 1;
+
+    IF NOT FOUND THEN RETURN; END IF; -- no submissions
+
+    IF check_submission_id IS NOT NULL and check_submission_id != submission.id THEN
+        RAISE EXCEPTION 'check_submission_id mismatch: % vs %', check_submission_id, submission.id;
+    END IF;
+
+    -- does the most recent submission actually need grading?
+    IF submission.broken THEN RETURN; END IF;
```

Comment:
Reaching into my lack of PL knowledge here--can a student still try to submit an answer to a broken question? Is there any disadvantage to allowing people to "manually grade" broken variants (either from generation/submit errors) as long as there's a pending job?

Reply:
I think the disadvantage of allowing people to "manually grade" broken variants is that there is more complexity. One would have to change the question to a manual type (if it isn't already). As we know, that's an issue in and of itself, as pre-existing instance questions do not take the latest updated configuration information. Further, one can assign a grade to a broken question through the UI already. 

My intuition is that a question that is broken but has a manual grade is invalid and should not be graded. It does not make sense to assign a manual grade to a broken question because it still uses the internal scoring mechanisms, which are based on an automatic portion. If that automatic portion is broken, then assigning the manual grade on top of 0 will result in a different score than if the question were not broken. 

I cannot recall if a student can submit an answer to a broken question, but I believe that the student can. I believe it simply creates another issue for the question when a broken question is detected.

Reply:
This makes sense to me!